packages/embedder/ai_models/RecursiveCharacterTextSplitter.ts

```diff
@@ -0,0 +1,117 @@
+// mostly borrowed from langchain
+export class RecursiveCharacterTextSplitter {
+  separators = ['\n\n', '\n', ' ', '']
+  keepSeparator = true
+  chunkSize: number
+  chunkOverlap: number
+  constructor({chunkSize, chunkOverlap}: {chunkSize: number; chunkOverlap: number}) {
+    this.chunkSize = chunkSize
+    this.chunkOverlap = chunkOverlap || 0
+  }
+  private splitOnSeparator(text: string, separator: string): string[] {
+    let splits: string[] = []
+    if (separator) {
+      if (this.keepSeparator) {
+        const regexEscapedSeparator = separator.replace(/[/\-\\^$*+?.()|[\]{}]/g, '\\$&')
+        splits = text.split(new RegExp(`(?=${regexEscapedSeparator})`))
+      } else {
+        splits = text.split(separator)
+      }
+    } else {
+      splits = text.split('')
+    }
+    return splits.filter((s) => s !== '')
+  }
+
+  private joinDocs(docs: string[], separator: string): string | null {
+    const text = docs.join(separator).trim()
+    return text === '' ? null : text
+  }
+
+  private mergeSplits(splits: string[], separator: string) {
+    const docs: string[] = []
+    const currentDoc: string[] = []
+    let total = 0
+    for (const d of splits) {
+      const _len = d.length
+      if (total + _len + (currentDoc.length > 0 ? separator.length : 0) > this.chunkSize) {
+        if (total > this.chunkSize) {
+          console.warn(
+            `Created a chunk of size ${total}, +
+which is longer than the specified ${this.chunkSize}`
+          )
+        }
+        if (currentDoc.length > 0) {
+          const doc = this.joinDocs(currentDoc, separator)
+          if (doc !== null) {
+            docs.push(doc)
+          }
+          // Keep on popping if:
+          // - we have a larger chunk than in the chunk overlap
+          // - or if we still have any chunks and the length is long
+          while (total > this.chunkOverlap || (total + _len > this.chunkSize && total > 0)) {
+            total -= currentDoc[0]!.length
+            currentDoc.shift()
+          }
+        }
+      }
+      currentDoc.push(d)
+      total += _len
```

Comment:
-1 I checked the python code you've linked and they add the separator length here as well which fixes the other places where I said there was a mismatch:

https://github.com/langchain-ai/langchain/blob/5d220975fc563a92f41aeb0907e8c3819da073f5/libs/text-splitters/langchain_text_splitters/base.py#L142

Reply:
whoaaa! good find. I'm surprised they don't have a test suite to make sure the JS result is the same as the python result. Would you like to PR langchainjs or shall I?

Reply:
https://github.com/langchain-ai/langchainjs/pull/4849